stdlib/stdlib.mli

```diff
@@ -51,10 +51,40 @@ exception Exit
 (** The [Exit] exception is not raised by any library function.  It is
     provided for use in your programs. *)
 
-val try_finally : (unit -> 'a) -> (unit -> unit) -> 'a
-(** [try_finally f g] calls [f ()] and returns its result.  If it raises,
-    the same exception is raised unless [g ()] itself raises;
-    in {b: any} case, [g ()] is called after [f ()] terminates.
+val try_finally :
+  ?always:(unit -> unit) ->
+  ?exceptionally:(unit -> unit) ->
+  (unit -> 'a) -> 'a
+(** [try_finally work ~always ~exceptionally] is designed to run code
+    in [work] that may fail with an exception, and has two kind of
+    cleanup routines:
+    {ul
+    {- [always], that must be run after {b any} execution
+       of the function (typically, freeing system resources), and}
+    {- [exceptionally], that should be run {b only} if [work] or [always]
```

Comment:
I'm not sold on this signature. I don't see any advantage in it, the primitive is bundling too many feature in my opinion. Somehow I find: 
```ocaml
try try_finally work ~finally:always with 
| e -> exceptionally (); raise e
```
to be perfectly fine and maybe even more clear; which is the semantics of your function if I understand correctly.

Reply:
I do agree with you. Keeping the simple version makes actually for clearer code, as you don’t need to carefully read the documentation to understand what is going on with exceptionally, and makes the implementation much simpler.

Reply:
The two versions are not equivalent when backtraces are enabled, if `exceptionally` terminates correctly but uses exceptions for internal control flow. On the other hand, the complexity reduction may still be worth it -- it's not like the rest of the stdlib is careful with backtraces anyway, and we could think of improving backtrace handling in the compiler/semantics instead of in the standard library.